Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -31,112 +31,143 @@ namespace firebase {
+ * assertion failure.
  *
- * @tparam T The contents of an element in the array.
+ * ArraySortedMap does not actually contain its array: it contains a shared_ptr
+ * to a FixedArray.
+ *
+ * @tparam T The type of an element in the array.
  * @tparam fixed_size the fixed size to use in creating the FixedArray.
  */
 template <typename T, ArraySortedMapBase::size_type fixed_size>
 class FixedArray {
  public:
-  typedef ArraySortedMapBase::size_type size_type;
-  typedef std::array<T, fixed_size> array_type;
-  typedef typename array_type::iterator iterator;
-  typedef typename array_type::const_iterator const_iterator;
+  using size_type = ArraySortedMapBase::size_type;
+  using array_type = std::array<T, fixed_size>;
+  using iterator = typename array_type::iterator;
+  using const_iterator = typename array_type::const_iterator;
+
+  FixedArray() : size_(0) {
+  }
+
+  template <typename SourceIterator>
+  FixedArray(SourceIterator src_begin, SourceIterator src_end) : FixedArray() {
+    append(src_begin, src_end);
+  }
 
   /**
    * Appends to this array, copying from the given src_begin up to but not
    * including the src_end.
    */
   template <typename SourceIterator>
   void append(SourceIterator src_begin, SourceIterator src_end) {
-    auto copy_end = std::copy(src_begin, src_end, end());
-    size_ = static_cast<size_type>(copy_end - begin());
+    size_type appending = static_cast<size_type>(src_end - src_begin);
+    size_type new_size = size_ + appending;
+    assert(new_size <= fixed_size);
+
+    std::copy(src_begin, src_end, end());
+    size_ = new_size;
   }
 
   /**
    * Appends a single value to the array.
    */
   void append(T&& value) {
-    *end() = std::forward<T>(value);
-    size_ += 1;
-  }
+    size_type new_size = size_ + 1;
+    assert(new_size <= fixed_size);
 
-  iterator begin() {
-    return contents_.begin();
+    *end() = std::move(value);
+    size_ = new_size;
   }
 
   const_iterator begin() const {
     return contents_.begin();
   }
 
-  iterator end() {
+  const_iterator end() const {
     return begin() + size_;
   }
 
-  const_iterator end() const {
-    return begin() + size_;
+  size_type size() const {
+    return size_;
   }
 
  private:
+  iterator begin() {
+    return contents_.begin();
+  }
+
+  iterator end() {
+    return begin() + size_;
+  }
+
   array_type contents_;
   size_type size_;
-
-  template <typename K, typename V, typename C>
-  friend class ArraySortedMap;
 };
 
+}  // namespace impl
+
 /**
  * ArraySortedMap is a value type containing a map. It is immutable, but has
  * methods to efficiently create new maps that are mutations of it.
  */
 template <typename K, typename V, typename C = std::less<K>>
-class ArraySortedMap : public ArraySortedMapBase {
+class ArraySortedMap : public impl::ArraySortedMapBase {
  public:
-  typedef ArraySortedMap<K, V, C> this_type;
-  typedef KeyComparator<K, V, C> key_comparator_type;
+  using this_type = ArraySortedMap<K, V, C>;
```

Comment:
Optional: for the class name itself, template parameter list may be omitted. It's called injected class name, see e.g. [this](https://stackoverflow.com/questions/25549652/why-is-there-an-injected-class-name).